Split standard library docs up one page per module.

sofiafaro-da opened the pull request:
Fixes #3172 

This PR updates our standard library docs to put each module on its own page. Users can easily navigate the docs by using the table of contents on the left, or the standard library page.

Here's a preview:

![image](https://user-images.githubusercontent.com/63245722/79769636-4a7d2c00-8324-11ea-8fe0-ee3d357d2067.png)

Summary of changes:

* Redid & pruned some bazel rules to get multiple outputs.
* Adjusted the daml-doc RST renderer output to support cross-reference links.
* Added a module index that just lists all the modules. This is also used to generate the navigation (automatically) in RST docs.
* Added an `--index-template` option to daml-docs, so you can have a separate template for the module index and individual modules. This was used to make the "standard library" page.
* Small annotation changes in the standard library to hide modules that should be hidden, and move other modules to the correct place.

TODO before merging:

*  [ ] Fix hoogle database URLs.

This change also highlights how a lot of our standard library documentation could be improved. Most modules don't yet have any description. Also, it might be better to put data types before typeclasses in each module doc. I will leave further documentation changes to a separate PR.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

cocreature commented on compiler/damlc/base-rst-index-template.rst:
```diff
@@ -0,0 +1,20 @@
+The standard library
+====================
+
+The DAML standard library is a collection of DAML modules that are bundled with the DAML SDK, and can be used to implement concrete DAML applications.
```
```suggestion
The DAML standard library is a collection of DAML modules that are bundled with the DAML SDK, and can be used to implement DAML applications.
```

cocreature commented on compiler/damlc/daml-prim-src/LibraryModules.daml:
```diff
@@ -5,7 +5,7 @@
 
 
--- Remove this module when DEL-6311 is done.
+-- | HIDE Remove this module when DEL-6311 is done.
```
Gotta love comments that refer to jira tickets that have been fixed.

hurryabit replied:
Can't we just remove this file?

cocreature replied:
Thinking about it, we fixed this for `damlc build` but we use `damlc compile` to compile daml-stdlib so it doesn’t help.

sofiafaro-da replied:
Probably ... I'm going to go through the stdlib in a separate PR to clean up docs and I'll try removing it then.

cocreature commented on compiler/damlc/daml-stdlib-src/DA/Internal/Compatible.daml:
```diff
@@ -4,7 +4,7 @@
 {-# LANGUAGE PatternSynonyms #-}
 
--- | Our Prelude, extending WiredIn with things that don't need special treatment.
+-- | HIDE Our Prelude, extending WiredIn with things that don't need special treatment.
```
This comment doesn’t make sense to me.

cocreature commented on compiler/damlc/daml-stdlib-src/DA/Generics.daml:
```diff
@@ -23,7 +23,7 @@
 {-# LANGUAGE DamlSyntax                 #-}
 
-
+-- | HIDE Partial implementation of DeriveGeneric. Unstable, use at great risk.
```
```suggestion
-- | HIDE Partial implementation of DeriveGeneric. Note that this is not supported in cross-SDK dependencies.
```
“use at great risk” sounds a bit too scary imho.

sofiafaro-da commented:
> Awesome, thanks! Do we need a redirect from the old docs url or does that just point to the main docs page that lists all modules?

That sounds like a good idea, I'll add that.

sofiafaro-da commented:
I made some more changes:

* Moved the standard library to `daml/stdlib` (instead of `daml/reference/base`), which reflects where it actually is in the docs (there's a whole history here that I don't want to get into ... but suffice it to say that this is where the standard library docs should have been before this PR, but weren't, but there's no longer a reason to have it this way).

* Updated the existing standard library redirects, and added a redirect from `daml/reference/base.html` to `daml/stdlib/index.html`.

* Changed the way modules are listed ... I removed the `Module ` prefix from the index / navigation because I think it looked a bit goofy.

* Updated the live preview script.

* Updated the hoogle base URL, the URL for generated links, and the builtin `damlBaseURL`. This is a little bit better than pointing to an old page, but it's still not correct in general, especially for Hoogle. The problem is that the way Hoogle is generated right now isn't very amenable to putting in the correct links.

If possible, I'll work on improving the Hoogle database problem in a separate PR. I don't want to make this one any larger than it already is... (Right now the behavior is that if you click on a hoogle link, it will take you to the standard library index instead of the exact module you want. That was only fine when everything was on the same page.)
